### PR TITLE
Error handling

### LINKED ISSUE #4956: Handling errors and other events
# Error handling in VSF Next
We will have a list of predefined errors and success messages for each composable.  We wanted to capture here a dynamic nature of such events and make it easy to use it both on specific components or via global notifications. We're not sure if we want to use both versions or one is enough (the wider API surface is the worse so we're trying to find justification for both APIs)

## User API
```ts
type AgnosticEvent = AgnosticError | AgnosticSuccess

//  {
//    type; 'error' | 'success',
//    id: string, // eg 'NOT_IN_STOCK'
//    message: string
//    $original: any // original error
//  }

const { cart, events } = useCart()

// you can listen to events happening
events.on(e => {
  if (e.type = 'error' && e.id = 'NOT_IN_STOCK') sendInAppNotification('error', e.message)
  if (e.type = 'success' && e.id = 'ADD_TO_CART') sendInAppNotification('success', e.message)
})

// or just use it as an object so you can use it in the different way and have typings
events.values //  { error: { NOT_IN_STOCK: errorVal, PRODUCT_DONT_EXIST: null, ... }, success { ... } }
```

### Some usage examples:
* Object version

```html
<template>
 <button @click="adddToCart(product)">Add to cart</button>
 <div v-if="cartEvents.error.notInStock">Product not in stock</dIv>
 <!-- you can also directly display the message from Error object but this doesn't mean you should :) -->
 <div v-if="cartEvents.error.productDontExist"> {{ cartEvents.error.productDontExist.message }}</dIv> 
</template>

<script>
export default {
  setup () {
    const { addToCart, events } = useCart()

    return {
      cartEvents: events.values
    }
  }
}
</script>
```

* Object version

```html
<template>
 <button @click="adddToCart(product)">Add to cart</button>
 <div v-if="error.type === et.NOT_IN_STOCK">Product not in stock</dIv> 
 <div v-if="error.type === et.PRODUCT_DONT_EXIST">{{ error.message }}</dIv>
</template>

<script>
export default {
  setup () {
    const { addToCart, events } = useCart()
    const error = ref(null)

    events.on(e => {
      if (e.type === 'error' ) error.value = e
    })

    return {
      et: events.error.types
      error
    }
  }
}
</script>
```

## Integrator API
```ts
const params: UseCartFactoryParams<Cart, LineItem, ProductVariant, any> = {
  // ...
  addToCart: async ({ currentCart, product, quantity }, { customQuery, report }) => {

    try {
      const { data } = await apiAddToCart(currentCart, product, quantity, customQuery);
      report.success.addToCart(product)
    } catch (e) {
      if (e.message === 'Not in stock') report.error.notInStock(e)
      report.error.custom('name', e)
    }
    return data.cart;
  },
}
```

┆Issue is synchronized with this [Jira Story](https://vsf.atlassian.net/browse/NEXT-317) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by Fifciu**
### Short Description and Why It's Useful

closes #4956

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: packages/core/core/src/factories/useCartFactory.ts [hunk 1]
```diff
@@ -43,55 +44,76 @@ export const useCartFactory = <CART, CART_ITEM, PRODUCT, COUPON>(
...
+        Logger.error('useCart/removeItem', err);
+      } finally {
         loading.value = false;
       }
     };
 
+    const updateItemQty = async ({ product, quantity, customQuery }) => {
+      Logger.debug('useCart.updateItemQty', { product, quantity });
+
+      if (quantity && quantity > 0) {
```
Comment (andrzejewsky) [inline packages/core/core/src/factories/useCartFactory.ts:94]: don't do it here, there will be some platforms where you won't even pass the quantity also it's something to handle on the UI

**Review by andrzejewsky — CHANGES_REQUESTED** [review]

File: packages/core/core/src/factories/useCartFactory.ts (hunk 1 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useCartFactory.ts:94]: We decided to leave it like it is for now.

File: packages/core/core/src/factories/useCartFactory.ts [hunk 2]
```diff
@@ -34,6 +34,7 @@ export const useCartFactory = <CART, CART_ITEM, PRODUCT, COUPON>(
     const loading: Ref<boolean> = sharedRef(false, 'useCart-loading');
     const cart: Ref<CART> = sharedRef(null, 'useCart-cart');
     const context = generateContext(factoryParams);
+    const error: Ref<ComposableErrors> = sharedRef({}, 'useCart-error');
```
Comment (filrak) [inline packages/core/core/src/factories/useCartFactory.ts:37]: IMHO this should be strictly typed per composable to allow autocompletion. I find this being one of the biggest values of this API, wdyt?

**Review by filrak — CHANGES_REQUESTED** [review]

File: packages/core/core/src/factories/useCartFactory.ts (hunk 2 above)
Comment (Fifciu) [reply packages/core/core/src/factories/useCartFactory.ts:37]: That's a good idea

**Review by filrak — APPROVED** [review]

**Review by filrak — CHANGES_REQUESTED** [review]
docs ;)

File: packages/core/core/src/types.ts [hunk 3]
```diff
@@ -216,26 +265,35 @@ export interface UseCheckout
...
 export interface UseReview<REVIEW, REVIEWS_SEARCH_PARAMS, REVIEW_ADD_PARAMS> {
   search(params: ComposableFunctionArgs<REVIEWS_SEARCH_PARAMS>): Promise<void>;
   addReview(params: ComposableFunctionArgs<REVIEW_ADD_PARAMS>): Promise<void>;
+  error: ComputedProperty<UseReviewComposableErrors>;
   reviews: ComputedProperty<REVIEW>;
   loading: ComputedProperty<boolean>;
   [x: string]: any;
 }
-
+export interface UseFacetComposableErrors {
```
Comment (filrak) [inline packages/core/core/src/types.ts:280]:
```suggestion
export interface UseFacetErrors {
```
imho without `Composable` its still perfectly self-descriptive and changing this is a simple s&r ;)

applies to all types

File: packages/core/docs/general/error-handling.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# Error handling
+
+A flexible way of error handling is essential for a framework like Vue Storefront. As factories of composables are hearth of our core - we decided to put there whole error handling mechanism.
```
Comment (filrak) [inline packages/core/docs/general/error-handling.md:3]: Factories are internal functionality. Most of the users reading the docs that never got that deep will not be aware of what they are and could get confused

**Review by filrak — CHANGES_REQUESTED** [review]

**Review by andrzejewsky — APPROVED** [review]

File: packages/core/docs/general/error-handling.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    return {
+      error
+    }
+  }
+}
+</script>
+```
+
+There is a dedicated interface for each composable, example one for the `useCart`:
```
Comment (filrak) [inline packages/core/docs/general/error-handling.md:27]:
```suggestion
There is a dedicated interface for each composable. Take a look at this one from the `useCart`:
```

**Review by filrak — APPROVED** [review]

**Pull request merged**